rule gatk4_genotype_gvcfs:
    input:
        ref=config["genome"],
        dict=re.sub("(\.fasta$|\.fa$)", ".dict", config["genome"]),
        gvcfs=config["vcfs"]["diploid"] + config["vcfs"]["tetraploid"]
    output:
        "data/calls/all-calls.vcf"
    threads:
        config['params']['gatk_genotype_gvcfs']['threads']
    log: "logs/gatk/gatk4-all-calls.log"
    shell:
        "gatk-launch GenotypeGVCFs "
        "-R {input.ref} "
        "$(echo {input.gvcfs} | sed 's/data/ --variant data/g') "
        "-O {output} "
        "2> {log}"
