rule gatk4_combine_sample_gvcfs:
    input:
        ref=config["genome"],
        dict=re.sub("(\.fasta$|\.fa$)", ".dict", config["genome"]),
        gvcfs=config["vcfs"]["diploid"] + config["vcfs"]["tetraploid"],
        intervals=config["intervals"]
    output:
        "data/calls/gatk_sample/all-calls.g.vcf"
    log:
        "logs/gatk/gatk4-all-calls.g.vcf.log"
    shell: """
        grep -i 'chr\|DM' {input.intervals} > tmp_chrom.intervals ;
        gatk-launch CombineGVCFs -R {input.ref} \
        $(echo {input.gvcfs} | sed 's/data/ --variant data/g') -O {output} \
        -L tmp_chrom.intervals 2> {log}
    """