# Generates fixed-size intervals for running GATK4 HaplotypeCaller in parallel on cluster
# Splits scaffolds, and is subject to minor edge effects.
# Updated version splits at scaffold joins, which are delineated by 50000 "N" nucleotides

rule gatk4_make_fixed_intervals:
    input:
        dict=re.sub("(\.fasta$|\.fa$)", ".dict", config["genome"]),
    output:
        "data/intervals/gatk-haplocaller.intervals"
    params:
        chunksize=config["params"]["make_intervals"]
    run:
        chunk_int = int(params.chunksize)
        ofh = output[0]
        o = open(ofh, 'w')
        with open(input.dict) as f:
            for line in f:
                if line.startswith("@SQ\tSN:"):
                    l = line.split('\t')
                    chrom = l[1].split(":")[1]
                    end = int(l[2].split(":")[1])
                    for i in range(1, end, chunk_int):
                        if i + chunk_int > end:
                            outstr=chrom + ":" + str(i) + "-" + str(end) + "\n"
                        else:
                            outstr=chrom + ":" + str(i) + "-" + str(i + chunk_int - 1) + "\n"
                        o.write(outstr)
        o.close()
