rule gatk3_genotype_gvcfs:
    input:
        ref=config["genome"],
        dict=re.sub("(\.fasta$|\.fa$)", ".dict", config["genome"]),
        gvcfs=config["vcfs"]["diploid"] + config["vcfs"]["tetraploid"]
    output:
        "data/calls/gatk/all-calls.vcf"
    threads: config["params"]["gatk_genotype_gvcfs"]["threads"]
    log: "logs/gatk/{output}.log"
    shell:
        "gatk -T GenotypeGVCFs "
        "-R {input.ref} "
        "$(echo {input.gvcfs} | sed 's/data/ --variant data/g') "
        "-nct {threads} "
        "-o {output} "
        "2> {log}"